### PR TITLE
Add GraphQL interface for global Artsy search

### PR DESCRIPTION
**Pull request by dblandin**
Alternate approach to https://github.com/artsy/metaphysics/pull/1475 which introduces a `Searchable` interface instead of defining a `SearchResult` type.

Add a new global search connection to support a reimplementation and redesign of the search bar and search results page on artsy.net.

This implementation introduces a new `Searchable` interface. Searchable entities represented in Metaphysics and supported by the Gravity API should implement this interface.

If entity-specific attributes are requested, the appropriate loader should be wired up so that the resolver can fetch the full object from Gravity and include it within the query response. We determine whether a
particular search result item should be fetched by inflecting on the GraphQL query AST and looking for an inline fragment matching the type of the search result item. If such an inline fragment is present, the
attributes provided by the `Searchable` interface alone may not satisfy the query.

**Example Queries**

Let's execute an `AUTOSUGGEST` search query, fetching the first 3 items of entity type `ARTIST` or `ARTWORK` based on the string "Bowi".

```graphql
{
  search(query: "Bowi", first: 3, entities: [ARTIST, ARTWORK], mode: AUTOSUGGEST) {
    edges {
      node {
        __typename
        displayLabel
        href
        imageUrl

        ... on SearchableItem {
          searchableType
        }
      }
    }
  }
}
```

```javascript
{
  "data": {
    "search": {
      "edges": [
        {
          "node": {
            "__typename": "SearchableItem",
            "displayLabel": "David Bowie",
            "href": "/artist/david-bowie",
            "imageUrl": "https://d32dm0rphc51dk.cloudfront.net/-c6JZBhhGEA0Rsdr3iyXLg/square.jpg",
            "searchableType": "Artist"
          }
        },
        {
          "node": {
            "__typename": "SearchableItem",
            "displayLabel": "Michi Broussard, BOWIE DIAMONDS - David Bowie (2018)",
            "href": "/artwork/5be09787783a6c00289d229f",
            "imageUrl": "https://d32dm0rphc51dk.cloudfront.net/7jIyO_OmESBkm9VGINxrDg/square.jpg",
            "searchableType": "Artwork"
          }
        },
        {
          "node": {
            "__typename": "SearchableItem",
            "displayLabel": "Michi Broussard, BOWIE - David Bowie (2018)",
            "href": "/artwork/5bcbca48bdc8dd5845c2928c",
            "imageUrl": "https://d32dm0rphc51dk.cloudfront.net/VzcDFWguDyEO6emGcM-EFA/square.jpg",
            "searchableType": "Artwork"
          }
        }
      ]
    }
  }
}
```

---

What if we want to execute the same query, but fetch additional Artist and Artwork-specific attributes as well? We can use inline fragments in our query.

Under the hood, this will trigger additional API requests for the entity types targeted by the inline fragments. See how the `__typename` changes from `SearchableItem` to `Artist` or `Artwork`.

```graphql
{
  search(query: "Bowie", first: 3, entities: [ARTIST, ARTWORK], mode: AUTOSUGGEST) {
    edges {
      node {
        __typename
        displayLabel
        href
        imageUrl

        ... on Artist {
          hometown
        }
        ... on Artwork {
          dimensions {
            in
            cm
          }
        }
      }
    }
  }
}
```

```javascript
{
  "data": {
    "search": {
      "totalCount": 3,
      "edges": [
        {
          "node": {
            "__typename": "Artist",
            "displayLabel": "David Bowie",
            "href": "/artist/david-bowie",
            "imageUrl": "https://d32dm0rphc51dk.cloudfront.net/-c6JZBhhGEA0Rsdr3iyXLg/square.jpg",
            "hometown": "Brixton, London, United Kingdom"
          }
        },
        {
          "node": {
            "__typename": "Artwork",
            "displayLabel": "BOWIE DIAMONDS - David Bowie",
            "href": "/artwork/michi-broussard-bowie-diamonds-david-bowie",
            "imageUrl": "https://d32dm0rphc51dk.cloudfront.net/7jIyO_OmESBkm9VGINxrDg/square.jpg",
            "dimensions": {
              "in": null,
              "cm": null
            }
          }
        },
        {
          "node": {
            "__typename": "Artwork",
            "displayLabel": "BOWIE - David Bowie",
            "href": "/artwork/michi-broussard-bowie-david-bowie",
            "imageUrl": "https://d32dm0rphc51dk.cloudfront.net/VzcDFWguDyEO6emGcM-EFA/square.jpg",
            "dimensions": {
              "in": "24 × 36 in",
              "cm": "61 × 91.4 cm"
            }
          }
        }
      ]
    }
  }
}
```

---

To execute a `SITE`-wide search (to be used eventually on the reimplemented / redesigned search results page), change the `mode` to `SITE`. You can also use inline fragements to fetch additional Artist and Artwork attributes with this search mode.

```graphql
{
  search(query: "Bowie", first: 3, entities: [ARTIST, ARTWORK], mode: SITE) {
    totalCount
    edges {
      node {
        __typename
        displayLabel
        href
        imageUrl
      }
    }
  }
}
```

```javascript
{
  "data": {
    "search": {
      "edges": [
        {
          "node": {
            "__typename": "SearchableItem",
            "displayLabel": "David Bowie",
            "href": "/artist/david-bowie",
            "imageUrl": "https://d32dm0rphc51dk.cloudfront.net/-c6JZBhhGEA0Rsdr3iyXLg/square.jpg",
            "searchableType": "Artist"
          }
        },
        {
          "node": {
            "__typename": "SearchableItem",
            "displayLabel": "Banz & Bowinkel",
            "href": "/artist/banz-and-bowinkel",
            "imageUrl": "https://d32dm0rphc51dk.cloudfront.net/Eja6bJ5dDp4tVRyJDx8NnA/square.jpg",
            "searchableType": "Artist"
          }
        },
        {
          "node": {
            "__typename": "SearchableItem",
            "displayLabel": "Wolf Böwig",
            "href": "/artist/wolf-bowig",
            "imageUrl": "https://d32dm0rphc51dk.cloudfront.net/tN4F_ZRQ4_ijtzmLJmCkOg/square.jpg",
            "searchableType": "Artist"
          }
        }
      ]
    }
  }
}
```

### CODE REVIEW

File: src/lib/loaders/loaders_with_authentication/search.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import factories from "../api"
+import * as url from "url"
+import { SearchIndex } from "schema/search"
+
+export const searchLoader = (accessToken, opts) => {
+  const gravityAccessTokenLoader = () => Promise.resolve(accessToken)
+  const { gravityLoaderWithAuthenticationFactory } = factories(opts)
+  const gravityLoader = gravityLoaderWithAuthenticationFactory(
+    gravityAccessTokenLoader
+  )
```
Comment (alloy) [inline src/lib/loaders/loaders_with_authentication/search.ts]: I think it's better to pass in a preconfigured `gravityLoader` from `src/lib/loaders/loaders_without_authentication/gravity.ts`, because in other cases [than gravity] a service loader will need to do some extra work to fetch a token from e.g. gravity, and we don't want to set an example where people may end up doing the wrong thing. (The wrong thing being to create a new service loader that refetches a token, rather than re-using the already fetched one.)

File: src/lib/loaders/loaders_without_authentication/search.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    }
+
+    switch (mode) {
+      case "AUTOSUGGEST":
+        return url.format({ pathname: "/match/suggest", query: queryParams })
+      default:
+        return url.format({ pathname: "/match", query: queryParams })
+    }
+  })
+}
```
Comment (alloy) [inline src/lib/loaders/loaders_without_authentication/search.ts]: I might be missing something, but it seems like this is identical to the authenticated one and so making the change to inject the already created `gravityLoader` would remove the need to have two copies of the `searchLoader`.

Also, we're following a file naming convention where the file is called the same as the main export, i.e. `searchLoader.ts`.

File: src/schema/search.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    SITE: {
+      value: "SITE",
+    },
+  },
+})
+
+export const searchArgs = {
+  query: {
+    type: GraphQLString,
+  },
```
Comment (alloy) [inline src/schema/search.ts]: I assume this one shouldn't be optional?

File: src/schema/search.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  indexes: {
+    type: new GraphQLList(SearchIndex),
+  },
+  mode: {
+    type: SearchMode,
+  },
```
Comment (alloy) [inline src/schema/search.ts]: I assume these are optional as the upstream service will have sane defaults? (A final version of the PR should include such defaults in the descriptions or the args.)

**Comment by alloy** [comment]
I figured it would be helpful to show examples of each PR.

```graphql
{
  search(query: "Banksy") {
    edges {
      node {
        # if you only want generic data
        ... on Searchable {
          label
        }
        # if you want specific data for each concrete type
        ... on Artist {
          name
        }
        ... on Artwork {
          title
        }
      }
    }
  }
}
```

**Comment by alloy** [comment]
After looking at both query examples, I tend to like this version better, because it would allow for more generic code on the client side when you have a component that shows e.g. a list of artworks, whereas with the other schema the component's fragment would have to be different to make the additional `item` selection. I'm sure that's relatively easy to overcome, but it's at least _a_ pro for this version.

File: src/schema/search.ts [hunk 5]
```diff
@@ -66,26 +63,85 @@ const hydrateSearchResult = (result, root) => {
...
+              ) {
+                shouldHydrate = true
+              }
+            },
+            FragmentSpread(_node) {
+              throw new Error(
+                "Named fragment spreads are currently unsupported for search."
+              )
+            },
+          })
```
Comment (alloy) [inline src/schema/search.ts]: We can probably return the BREAK constant here to stop further recursion of the AST

File: src/schema/search.ts [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    visit(info.fieldNodes[0], {
+      Field(node) {
+        if (node.name.value === "node") {
+          visit(node, {
+            InlineFragment(node) {
+              if (
+                node.typeCondition &&
+                node.typeCondition.name.value !== SearchableItem.name
+              ) {
+                shouldHydrate = true
```
Comment (alloy) [inline src/schema/search.ts]: We can probably return the BREAK constant here to stop further recursion of the AST

**Comment by alloy** [comment]
I just realized that we could have moved the hydration into the `node` field resolver and have to do less visiting. Additionally I think the thrown error would be better understandable as its metadata would point to the `node` field being the origin. Anyways, we can follow up after deciding what schema to use.

File: src/lib/loaders/searchLoader.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
+// @ts-check
```
Comment (damassi) [inline src/lib/loaders/searchLoader.ts]: Is this needed?

Comment (dblandin) [reply src/lib/loaders/searchLoader.ts]: Probably not!

**Comment by dblandin** [comment]
> I just realized that we could have moved the hydration into the node field resolver and have to do less visiting. Additionally I think the thrown error would be better understandable as its metadata would point to the node field being the origin. Anyways, we can follow up after deciding what schema to use.

@alloy I'm not entirely sure where to make that change. Could you add a code snippet or maybe we can chat tomorrow?

File: src/schema/search.ts [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+export const searchArgs = pageable({
+  query: {
+    type: new GraphQLNonNull(GraphQLString),
+    description: "Search query to perform. Required.",
+  },
+  indexes: {
+    type: new GraphQLList(SearchIndex),
+    description: "Indexes to include in search. Default: [ARTIST, ARTWORK].",
+  },
```
Comment (dblandin) [inline src/schema/search.ts]: How do we feel about this argument name? Does it reveal too much of underlying system? Maybe we could rename this to `entities`?

File: src/lib/loaders/loaders_with_authentication/gravity.ts [hunk 9]
```diff
@@ -72,6 +73,7 @@ export default (accessToken, userID, opts) => {
       "is_saved"
     ),
     savedArtworksLoader: gravityLoader("collection/saved-artwork/artworks", { user_id: userID, private: true }),
+    searchLoader: searchLoader(gravityLoader),
```
Comment (mzikherman) [inline src/lib/loaders/loaders_with_authentication/gravity.ts]: Not sure that we even need this authenticated loader (search shouldn't really change based on your access/identification IMO, i wouldnt think).

Comment (dblandin) [reply src/lib/loaders/loaders_with_authentication/gravity.ts]: Yep, looks like there isn't any difference :+1:

File: src/schema/search.ts (hunk 8 above)
Comment (mzikherman) [reply src/schema/search.ts]: I'm ok with it...but I see your point. 🤔 will come back if something comes to mind, but I don't feel strongly!

File: src/schema/search.ts [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+      ...object,
+      __typename: result.label,
+    }
+  })
+}
+
+const shouldHydrate = info => {
+  let shouldHydrate = false
+
+  visit(info.fieldNodes[0], {
```
Comment (mzikherman) [inline src/schema/search.ts]: This block is intense! I'd love a comment describing what's going on.

File: src/schema/search.ts [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+    const gravityArgs = {
+      ...pageOptions,
+      indexes: args.indexes,
+      total_count: true,
+    }
+
+    const hydrate = shouldHydrate(info)
+
+    return source.searchLoader(gravityArgs).then(({ body, headers }) => {
+      const totalCount = parseInt(headers["x-total-count"])
```
Comment (mzikherman) [inline src/schema/search.ts]: i dont think the `/match` and `/match/suggest` endpoints support returning counts. The count being placed in the header only happens 'for free' if you're using the conventional `paginate` helper in the API: https://github.com/artsy/gravity/blob/a9c7e6f229ad463bb5b8a00f38e0f65ebf1d6e78/app/api/util/page_helpers.rb#L30

Since the search endpoint returns an array of results, I don't think we can modify that payload, so we'll need to add the total into the header ourselves.

Basically, we need: `header 'X-Total-Count', results['hits']['total'].to_s if params[:total_count]` added in the endpoint: https://github.com/artsy/gravity/blob/6f831b886683f432d31a52cbcc893f68aac47943/app/api/v1/match_endpoint.rb#L195

File: src/schema/search.ts [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+    type: new GraphQLList(SearchIndex),
+    description: "Indexes to include in search. Default: [ARTIST, ARTWORK].",
+  },
+  mode: {
+    type: SearchMode,
+    description: "Mode of search to exceute. Default: SITE.",
+  },
+})
+
+const hydrateSearchResult = (result, root) => {
```
Comment (mzikherman) [inline src/schema/search.ts]: An enhancement (not for now! but a nice follow up from @zephraph work), is that we should be batch hydrating results. If we get a list of 10 artists back w/ their partial information from ES, and want to hydrate them, rather than 10 individual fetches for each artist, one batched one would be nice.

I think that's an enhancement and not a blocker b/c we can see if it's even needed (perf might be acceptable w/o the batching enhancement), and if the perf is an issue we can always just skip hydration and use the `SearchableItem` subset of fields (and can even enhance those from Gravity if absolutely needed).

File: src/schema/search.ts [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+              __typename: "SearchableItem",
+            })
+          }
+        })
+      ).then(results => {
+        const connection = connectionFromArraySlice(results, args, {
+          arrayLength: totalCount,
+          sliceStart: pageOptions.offset,
+        })
+        // TODO: We shouldn't have to override the pageInfo here
```
Comment (mzikherman) [inline src/schema/search.ts]: I don't fully follow the comment, but I think you do need to override/merge the modified implementations for `hasPreviousPage` and `hasNextPage`, so this looks good to me.

The way those fields are calculated by the lib we're using doesn't fully work. Here's that implementation: https://github.com/graphql/graphql-relay-js/blob/582b0162709bd2db45cbd3925e2e69d9d269ec7e/src/connection/arrayconnection.js#L106-L108 

As you're scrolling forward ', by using `(first: ..., after: ...)` as your connection paging arguments, it can't determine if there are previous pages (and vice versa), so the lib returns `false` (kind of a weird return). So this slightly laborious construction of the connection return type (cursors, count, connection, page info), seems 👍 . We do this in several places so it's worth refactoring or extracting that into a wrapper of some kind.

File: src/schema/searchableItem.ts [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+import { Searchable } from "schema/searchable"
+import { NodeInterface } from "schema/object_identification"
+
+export const SearchableItem = new GraphQLObjectType({
+  name: "SearchableItem",
+  interfaces: [NodeInterface, Searchable],
+  fields: {
+    __id: {
+      type: new GraphQLNonNull(GraphQLID),
+      resolve: item => toGlobalId("SearchableItem", item._id),
```
Comment (mzikherman) [inline src/schema/searchableItem.ts]: nice

File: src/schema/search.ts [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+
+import { Searchable } from "./searchable"
+import { convertConnectionArgsToGravityArgs } from "lib/helpers"
+
+import { visit, BREAK } from "graphql"
+
+export const SearchIndex = new GraphQLEnumType({
+  name: "SearchIndexes",
+  values: {
+    ARTWORK: {
```
Comment (mzikherman) [inline src/schema/search.ts]: We can add `MarketingCollection` at some point 😎 

That's going to be...interesting since thats merged in from KAWS. 👻 

The good news is (not only should we save that for a separate PR), but if there's some issues w/ that, we can skip rehydration for those types and just rely on the `... on SearchableItem` fields. Currently the ES data for collections just contains the text we're interested in searching, but it's easy to populate the image there if needed. I don't think there's any other data the collection offers that we even need (besides the title/label/image/href), so that's why I'm thinking we could just wind up skipping hydration for those types.

**Comment by mzikherman** [comment]
Amazing and impressive PR 👏 

Overall looks great, once there's specs and you're satisfied things are 'working' as best you can tell and things are just feeling reasonable in GraphiQL pointing to staging, feel free to merge! We'll be able to correct any issues easily since no one will be using it yet anyway 😜

File: src/schema/search.ts (hunk 13 above)
Comment (alloy) [reply src/schema/search.ts]: When we stepped into the lib code it seemed mostly due to a missing `last` (page number) parameter, but for some reason the code that determined `hasPreviousPage` would still result in `false` for reasons that weren't immediately clear. In short, this is a comment to look into that again at some point and clean all these cases up, as it should be able to work if we pass in the right parameters.

File: src/schema/search.ts (hunk 11 above)
Comment (alloy) [reply src/schema/search.ts]: There was definitely a value ( 395 iirc) coming back, although it wasn’t immediately clear where in Gravity’s endpoint that was coming back from.

Comment (mzikherman) [reply src/schema/search.ts]: yea, that's present in the main `/match` endpoint: https://github.com/artsy/gravity/blob/a9c7e6f229ad463bb5b8a00f38e0f65ebf1d6e78/app/api/v1/match_endpoint.rb#L178

aka `SITESEARCH`

but it's not present in the `/match/suggest` endpoint, which is where we need it for the new Autosuggest dropdown.

File: src/schema/search.ts (hunk 13 above)
Comment (mzikherman) [reply src/schema/search.ts]: Well, those are missing b/c that's actually passed in by the caller. You have two options for scrolling: 'forward' which means providing `(first: ..., after: ....`), or 'backward' which means providing `(last: ...., before: ...)`. Theoretically those are mutually exclusive (but we don't do that validation).

Since we just do regular `page/size` and windowed pagination, we _always_ paginate those via `(first: ..., after: ...)`, and so that's what requires the manual `hasXPage` calculation.
